### DISCUSSION TITLE
redacting-sensitive-data-with-microsoft-extensions-compliance/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#401] [Announcements]
# redacting-sensitive-data-with-microsoft-extensions-compliance/

In this post I show how you can add redaction to your [LogProperties] and [TagProvider] objects using the Microsoft.Extensions.Compliance.Redaction package

https://andrewlock.net/redacting-sensitive-data-with-microsoft-extensions-compliance/

**Comment by joperezr** [comment]
Thanks a bunch for writing this, Andrew. Just a small comment around:

> Note that you must apply the data classifications here if you're using [TagProvider]. Tags added using [TagProvider] ignore the [PersonalData] and [SensitiveData] attributes applied to the WeatherForecast object.

We are planning on adding an analyzer that detects this and flags it, similar to how `RequiresUnreferencedCode` attribute operates. This will be very important in order to guarantee that classified data doesn’t unintentionally _escape_ its classification.

**Reply by andrewlock** [reply]
Sounds great! I agree, that will be important, and is a nice solution 🙂

**Comment by joperezr** [comment]
> To be honest, I think defaulting to the NullRedactor is a mistake here. A better option IMO would be to not register a default, and to just throw. This would give guidance about what's missing. Unfortunately, I think that ship has likely sailed.

This is great feedback! I agree with you that this can be very confusing, especially when trying it for the first time. We should have an issue in the repo tracking this to explore options. Let me know if you are planning to log one, otherwise I can do it.

**Comment by sander1095** [comment]
Hi Andrew! Many thanks for these posts on these logging methods, they are extremely useful.

I'd like to ask how and where you learned about this. I consider myself quite up to date with updates in .NET land, but I had never heard about this before your blog post. And that's precisely why I find your posts *so* valuable!

**Reply by andrewlock** [reply]
Thanks, glad you find them useful! 🙂

I mentioned where I first heard about them in https://andrewlock.net/behind-logproperties-and-the-new-telemetry-logging-source-generator/#logging-whole-objects-with-logproperties-:

As part of the .NET 8 launch at .NET Conf 2023, Microsoft announced several new packages based on adding telemetry, observability, and resilience to your applications. The `[LogProperties]` attribute was [a relatively minor announcement](https://www.youtube.com/live/vU-iZcxbDUk?si=ekHW7csppCc2NJIs&t=3035) as part of general additional ILogger features added in .NET 8.

I was interested in learning more about exactly how it works, so I went digging through the repositories to see how they had implemented things. Essentially I searched the dotnet repos for `LogPropertiesAttribute` and from there I discovered this gold mine of extra functionality that wasn't touched on at all as far as I'm aware! 😃

**Comment by FirePT** [comment]
Hi Andrew, great article shining some light on this library, I've been implementing these new features in my current project and found some behaviors that need some sharing in my opinion. When you try to develop your own Custom Redactor you will be faced with some content added to the final redaction value because of an implementation detail regarding security. You can read all about it here : https://github.com/dotnet/extensions/issues/4836

The other challenge I encountered involved integrating the Serilog library with the Redaction extension. To seamlessly achieve this, it is essential to instantiate the Serilog instance, configure it, and then inject it into the WebApplication.CreateBuilder(args).Services.AddLogging() as illustrated below:
```
Log.Logger = new LoggerConfiguration()
    .Enrich.FromLogContext()
    .WriteTo.Console();

builder.Services.AddLogging(opts =>
{
    opts.AddSerilog(Log.Logger, dispose: true);
    opts.EnableRedaction();
});
```
I hope this saves someone some time digging in the documentation.

**Reply by TejendraPrasad** [reply]
Hi Andrew, thank you great article, I have been trying this in sample project :)

As FirePT has mentioned, i am still struggling to integrate with SeriLog...

Below is my current Serilog setup...

builder.Host.UseSerilog((context, loggerConfig) =>
    loggerConfig.ReadFrom.Configuration(context.Configuration));

Is there any sample to find out how to set up with Serilog... 

Thank you :)

**Comment by davardoust** [comment]
Hi Andrew! I truly appreciate your informative posts on logging techniques. 

I'm wondering if using `LoggerMessage` is mandatory. If a user chooses to log a specific property like `logger.log(f.TemperatureC)` instead of using the provided method `logger.LogWeatherForecast(f)`, I've noticed it doesn't function correctly. Is there a way to make the logger work properly in such cases without relying on the `LoggerMessage` class?

**Reply by joperezr** [reply]
Added a reply to your comment on my gist. Essentially the redaction functionality is not tied at all to the `LoggerMessage` generator, and can be used standalone. If you take a look at the generated method from the generator, you'll find that internally it is just getting a redactor from the DI container and then calling Redact on the message. If you are calling `logger.log` directly, you'd have to fetch the Redactor and call Redact on it yourself before the message is logged.

**Comment by dariogriffo** [comment]
Hi Andrew, thanks for the step by step example.
Have you manage to make this work with nested properties?
```csharp
public class A
{
    [SensitiveData]public strnig EmailAddress{get;set;}
    public B B{get;set;}
}

public class B
{
    [SensitiveData]public strnig Foo{get;set;}
}
```

With this classes I only get redacted the properties from A.

**Comment by AESSoftware** [comment]
In later posts we'll see how you can enable redaction in other parts of the framework

Any progress on that?
